rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+use core::ffi::c_void;
+use std::mem;
+use std::ptr;
+use std::vec;
+use symcrypt_sys;
+use crate::errors::SymCryptError;
+
+pub struct GcmState {
+    expanded_key: symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY,
+    state: symcrypt_sys::SYMCRYPT_GCM_STATE,
+}
+
+impl GcmState {
+    pub fn new(key: &[u8], nonce: &[u8]) -> Result<Box<Self>, SymCryptError> {
+        let mut instance = Box::new(GcmState {
```

Comment:
Have to `Box<>` since memory address for Self is moved around when returning. My guess is that it has to do with `Result<>` wrapping the returned values. When I did this code without a `Result<>` the code worked fine. To be safe, we should go back and modify other symcrypt structs to return a `Box<>`'d version. Created an issue to track this.

Reply:
The reason that memory address cannot change is because of underlying SymCrypt code. SymCrypt's magic checks require that function has the exact pointer. [Example](https://github.com/microsoft/SymCrypt/blob/main/lib/gcm.c#L334C1-L335C1)

Reply:
Please add your comments here  to the code.

Reply:
Added

Reply:
This is something I chatted with @nnmkhang a bit offline about, and doing this perfectly right is going to be a little tricky. Returning a `Box` will work most of the time, but people will still be able get into trouble if they try hard enough. For example, they could do:

```rust 
let state: Box<GcmState> = GcmState::new()?;
let new_state = Box::new(*state);
```

This moves the previously created state into a new location in memory, which means the magic value will be wrong again.

To do this fully correctly, you'd need to use [`Pin`](https://doc.rust-lang.org/std/pin/index.html). Unfortunately, `Pin` is really hard to use.

I think using `Box` is reasonable for the sake of ease of use. As I understand it, we don't get undefined behavior if the magic is wrong, but the functions that check the magic will either fail or shutdown the program. If that's the case, I think it'd be fine to note that in the doc comments and maybe provide a function that updates the magic that you can call if you really need to move out of a box for some reason.

If you wanted to go the `Pin` route, I'd recommend trying to keep that out of the public API as much as possible. One way you could do that is to build your own wrapper types that handle pinning internally. I'm not 100% sure what this would look like though, so it'd need a little more design work. This might be a good idea even for boxes though, because you could prevent some of the cases where people can get into trouble, like the example I gave.

Reply:
Yes, Box is ensuring that only one instance of the object is allocated on the heap.